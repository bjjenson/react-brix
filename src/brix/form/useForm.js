import { Map } from 'immutable'
import { Form } from './Form'
import { createReducer, generateDefaultFieldState } from './createReducer'
import { actions } from './fieldReducer'
import {
  useBooleanField,
  useNumberField,
  useSelectField,
  useTextField,
} from './fields'

/**
 * @param  param0 { import("../form").IFormProps }
 */
export const useForm = ({ fields, submit, validate, options = {}, initialValues = Map() }) => {
  const [state, dispatch] = createReducer({ fields, initialValues, options })

  const fieldData = state.keySeq().reduce((acc, fieldName) => {
    const fieldState = state.get(fieldName)
      .set('getAllValues', () => getFieldValues(fieldData))

    const fieldType = state.getIn([fieldName, 'initial', 'type'])
    const field = state.getIn([fieldName, 'initial', 'field'])
    switch (fieldType) {
      case 'select':
        acc[fieldName] = useSelectField(fieldState, dispatch, field)
        break
      case 'boolean':
        acc[fieldName] = useBooleanField(fieldState, dispatch, field)
        break
      case 'number':
        acc[fieldName] = useNumberField(fieldState, dispatch, field)
        break
      case 'text':
      default:
        acc[fieldName] = useTextField(fieldState, dispatch, field)
        break
    }
    return acc
  }, {})

  const addField = field => {
    const fieldState = generateDefaultFieldState(field, initialValues)
    dispatch(actions.insertField(field.name, fieldState))
  }

  const removeField = fieldName => {
    dispatch(actions.removeField(fieldName))
  }

  const tryValidateForm = () => {
    if (validate) {
      const values = getFieldValues(fieldData)
      return validate(values)
    }

    return {}
  }

  const trySubmitForm = () => {
    let isFormValid = true
    const formResults = tryValidateForm()
    const values = Object.entries(fieldData).reduce((acc, [key, v]) => {
      if (formResults[key]) {
        v.setValidationResult({ error: true, helperText: formResults[key] })
        isFormValid = false
      }

      const isFieldValid = v.validate()
      isFormValid = isFormValid && isFieldValid

      return acc.set(key, v.props.value)
    }, initialValues)
    if (isFormValid) {
      submit(values)
    }
  }

  const setValue = (fieldName, value) => {
    fieldData[fieldName].setValue(value)
  }

  return {
    ...getFieldProps(fieldData),
    setValue,
    submit: trySubmitForm,
    Form,
    addField,
    removeField,
  }
}

const getFieldValues = (fieldData) => {
  return Object.entries(fieldData).reduce((acc, [k, v]) => {
    acc[k] = v.props.value
    return acc
  }, {})
}

const getFieldProps = (fieldData) => {
  return Object.entries(fieldData).reduce((acc, [k, v]) => {
    acc[k] = v.props
    return acc
  }, {})
}
