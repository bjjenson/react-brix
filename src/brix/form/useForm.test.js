import { useState } from 'react'
import { fromJS } from 'immutable'
import { useForm } from './useForm'

jest.mock('react')

const setState = jest.fn()
const submitWorker = jest.fn()
let formArgs
beforeEach(() => {
  formArgs = [
    { name: 'name', type: 'text', label: 'The name' },
    { name: 'phone', type: 'text', label: 'The phone' },
  ]

  useState.mockImplementation(a => [a, setState])
})

test('returns array of form props', () => {
  const actual = useForm({ fields: formArgs })
  expect(actual).toMatchSnapshot()
})

test('throws if name is not set', () => {
  delete formArgs[0].name
  expect(() => useForm({ fields: formArgs })).toThrow()
})

test('no submit if field validation fails', () => {
  const validateName = jest.fn()
  validateName.mockReturnValue('name has error')
  formArgs[0].validate = validateName

  const actual = useForm({ fields: formArgs, worker: submitWorker })
  actual.submit()
  expect(submitWorker).not.toHaveBeenCalled()
})

test('no submit if form validation fails', () => {
  const validateForm = jest.fn()
  validateForm.mockReturnValue({ name: 'i am error' })
  const actual = useForm({ fields: formArgs, worker: submitWorker, validate: validateForm })
  actual.submit()
  expect(validateForm.mock.calls[0]).toMatchSnapshot()
  expect(submitWorker).not.toHaveBeenCalled()
})

test('submit calls worker, merges with initialValues', () => {
  formArgs = [
    { name: 'name' },
    { name: 'phone', value: 'b' },
  ]
  const initialValues = fromJS({
    name: 'a',
    id: 'id',
  })
  const actual = useForm({ fields: formArgs, submit: submitWorker, initialValues })
  actual.submit()
  expect(submitWorker.mock.calls[0]).toMatchSnapshot()
})
