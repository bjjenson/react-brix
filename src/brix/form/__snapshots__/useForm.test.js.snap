// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no submit if form validation fails 1`] = `
Array [
  Object {
    "name": "",
    "phone": "",
  },
]
`;

exports[`returns array of form props 1`] = `
Object {
  "Form": [Function],
  "addField": [Function],
  "name": Object {
    "label": Immutable.Map {
      "value": "current value of name",
      "getAllValues": [Function],
    },
    "value": "",
  },
  "phone": Object {
    "label": Immutable.Map {
      "value": "current value of phone",
      "getAllValues": [Function],
    },
    "value": "",
  },
  "removeField": [Function],
  "setValue": [Function],
  "submit": [Function],
}
`;

exports[`submit calls worker, merges with initialValues 1`] = `
Array [
  Immutable.Map {
    "name": "new name",
    "id": "id",
    "phone": "new phone",
  },
]
`;
